source/stream.js

```diff
@@ -0,0 +1,62 @@
+import {isReadableStream} from 'is-stream';
+
+export const getAsyncIterable = stream => {
+	if (isReadableStream(stream, {checkOpen: false})) {
+		return getStreamIterable(stream);
+	}
+
+	if (stream?.[Symbol.asyncIterator] === undefined) {
+		throw new Error('The first argument must be a Readable, a ReadableStream, or an async iterable.');
+	}
+
+	return stream;
+};
+
+// The default iterable for Node.js streams does not allow for multiple readers at once, so we re-implement it
+const getStreamIterable = async function * (stream) {
+	if (nodeImports === undefined) {
+		await loadNodeImports();
+	}
+
+	const controller = new AbortController();
+	handleStreamEnd(stream, controller);
+
+	try {
+		for await (const [chunk] of nodeImports.events.on(stream, 'data', {
+			signal: controller.signal,
+			highWatermark: stream.readableHighWaterMark,
+		})) {
+			yield chunk;
+		}
+	} catch (error) {
+		if (!controller.signal.aborted || error.cause === undefined) {
+			throw error;
+		} else if (error.cause !== abortError) {
+			throw error.cause;
+		}
+	} finally {
+		stream.destroy();
+	}
+};
+
+const handleStreamEnd = async (stream, controller) => {
+	try {
+		await nodeImports.streamPromises.finished(stream, {cleanup: true, readable: true, writable: false, error: false});
+		controller.abort(abortError);
+	} catch (error) {
+		controller.abort(error);
+	}
+};
+
+const abortError = new Error('Internal error');
+
+// Use dynamic imports to support browsers
+const loadNodeImports = async () => {
+	const [events, streamPromises] = await Promise.all([
+		import('node:events'),
+		import('node:stream/promises'),
+	]);
+	nodeImports = {events, streamPromises};
+};
```

Comment:
The `on()` and `finished()` functions have all the logic we need for this PR, and it does not quite make sense to re-implement them.

However, browsers don't have those methods. They also don't have `Readable` streams, so cannot run this part of the logic.

This PR is currently fixing this using dynamic imports. This is not great. Besides the hacky feel to this, it's got one practical issue which is: when executing `get-stream`, users might assume that the stream `data` and `error` events are being listened to as soon as `getStream()` starts. But this those dynamic imports, that's not the case. This means if another consumer has called `stream.on('data')`, it might start pulling some data that `getStream()` might miss. This also means any `error` event on the stream will be unhandled (which crashes the process) while the dynamic imports are resolving.

Reply:
We might also combine the last two solutions:
  - We use static `import ... from 'node:stream'`
  - We provide a second entrypoint without those static imports for browser users whose bundler cannot handle `node:*` imports

Reply:
Unlikely scenario, but I think using separate browser and node exports in package.json may be the best solution in this case.

Reply:
Sounds good. 
Should I do this in a second PR after this one?

Reply:
👍